File: Samples for Parallel Programming with the .NET Framework/[C#][VB][C++][F#]-Samples for Parallel Programming with the .NET Framework/C# and C++ and F# and VB/C#,C++,F#,VB/ShakespeareanMonkeys/ShakespeareanMonkeys_CSharp/MainForm.cs
//--------------------------------------------------------------------------
// 
//  Copyright (c) Microsoft Corporation.  All rights reserved. 
// 
//  File: MainForm.cs
//
//--------------------------------------------------------------------------

using System;
using System.Drawing;
using System.Threading;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace ShakespeareanMonkeys
{
    public partial class MainForm : Form
    {
        private static string _targetText =
            "To be or not to be, that is the question;" + Environment.NewLine +
            "Whether 'tis nobler in the mind to suffer" + Environment.NewLine +
            "The slings and arrows of outrageous fortune," + Environment.NewLine +
            "Or to take arms against a sea of troubles," + Environment.NewLine +
            "And by opposing, end them.";
        private TaskFactory _uiTasks;

        public MainForm()
        {
            InitializeComponent();

            txtTarget.Text = _targetText;
            _uiTasks = new TaskFactory(TaskScheduler.FromCurrentSynchronizationContext());
        }

        private int _currentIteration;
        private CancellationTokenSource _cancellation;

        private void btnRun_Click(object sender, EventArgs e)
        {
            if (_cancellation == null)
            {
                _cancellation = new CancellationTokenSource();
                GeneticAlgorithmSettings settings = new GeneticAlgorithmSettings { PopulationSize = Int32.Parse(txtMonkeysPerGeneration.Text) };

                txtBestMatch.BackColor = SystemColors.Window;
                lblGenerations.BackColor = SystemColors.Control;
                lblGenPerSec.Text = lblGenerations.Text = "-";
                lblElapsedTime.Text = "0";
                btnRun.Text = "Cancel";
                chkParallel.Visible = false;

                _startTime = _lastTime = DateTimeOffset.Now;
                timerElapsedTime.Start();

                // Run the work in the background
                _cancellation = new CancellationTokenSource();
                var token = _cancellation.Token;
                bool runParallel = chkParallel.Checked;
                Task.Factory.StartNew(() =>
                {
                    // Create the new genetic algorithm
                    var ga = new TextMatchGeneticAlgorithm(runParallel, _targetText, settings);
                    TextMatchGenome? bestGenome = null;

                    // Iterate until a solution is found or until cancellation is requested
                    for (_currentIteration = 1; ; _currentIteration++)
                    {
                        token.ThrowIfCancellationRequested();

                        // Move to the next generation
                        ga.MoveNext();

                        // If we've found the best solution thus far, update the UI
                        if (bestGenome == null ||
                            ga.CurrentBest.Fitness < bestGenome.Value.Fitness)
                        {
                            bestGenome = ga.CurrentBest;
                            _uiTasks.StartNew(() => txtBestMatch.Text = bestGenome.Value.Text);

                            // If we've found the solution, bail.
                            if (bestGenome.Value.Text == _targetText) break;
                        }
                    }

                    // When the task completes, update the UI
                }, token).ContinueWith(t =>
                {
                    timerElapsedTime.Stop();
                    chkParallel.Visible = true;
                    btnRun.Text = "Start";
                    _cancellation = null;

                    switch (t.Status)
                    {
                        case TaskStatus.Faulted:
                            MessageBox.Show(this, t.Exception.ToString(), "Error");
                            break;
                        case TaskStatus.RanToCompletion:
                            txtBestMatch.BackColor = Color.LightGreen;
                            lblGenerations.BackColor = Color.LemonChiffon;
                            break;
                    }
                }, _uiTasks.Scheduler);
            }
            else _cancellation.Cancel();
        }

        private DateTimeOffset _startTime = DateTimeOffset.MinValue;
        private DateTimeOffset _lastTime = DateTimeOffset.MinValue;

        private void timerElapsedTime_Tick(object sender, EventArgs e)
        {
            var now = DateTimeOffset.Now;
            var elapsed = (int)(now - _startTime).TotalSeconds;
            
            lblElapsedTime.Text = elapsed.ToString();
            lblGenerations.Text = _currentIteration.ToString();

            if (elapsed > 2)
            {
                var diffSeconds = (now - _lastTime).TotalSeconds;
                if (diffSeconds > 0)
                {
                    lblGenPerSec.Text = ((int)(_currentIteration / diffSeconds)).ToString();
                }
            }
        }
    }
}

File: Samples for Parallel Programming with the .NET Framework/[C#][VB][C++][F#]-Samples for Parallel Programming with the .NET Framework/C# and C++ and F# and VB/C#,C++,F#,VB/ShakespeareanMonkeys/ShakespeareanMonkeys_VisualBasic/MainForm.vb
'--------------------------------------------------------------------------
' 
'  Copyright (c) Microsoft Corporation.  All rights reserved. 
' 
'  File: MainForm.vb
'
'--------------------------------------------------------------------------

Imports System.Threading
Imports System.Threading.Tasks
Imports System.Windows.Forms
Imports System.Text

Namespace ShakespeareanMonkeys
    Partial Public Class MainForm
        Inherits Form

        Private Shared _targetText As String =
         "To be or not to be, that is the question;" & Environment.NewLine &
         "Whether 'tis nobler in the mind to suffer" & Environment.NewLine &
         "The slings and arrows of outrageous fortune," & Environment.NewLine & "Or to take arms against a sea of troubles," &
         Environment.NewLine & "And by opposing, end them."
        Private _uiTasks As TaskFactory

        Public Sub New()
            InitializeComponent()

            txtTarget.Text = _targetText
            _uiTasks = New TaskFactory(TaskScheduler.FromCurrentSynchronizationContext())
        End Sub

        Private _currentIteration As Integer
        Private _cancellation As CancellationTokenSource

        Private Sub btnRun_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnRun.Click
            If _cancellation Is Nothing Then
                _cancellation = New CancellationTokenSource()
                Dim settings = New GeneticAlgorithmSettings With {.PopulationSize = Int32.Parse(txtMonkeysPerGeneration.Text)}

                txtBestMatch.BackColor = SystemColors.Window
                lblGenerations.BackColor = SystemColors.Control
                lblGenerations.Text = "-"
                lblGenPerSec.Text = lblGenerations.Text
                lblElapsedTime.Text = "0"
                btnRun.Text = "Cancel"
                chkParallel.Visible = False

                _lastTime = DateTimeOffset.Now
                _startTime = _lastTime
                timerElapsedTime.Start()

                ' Run the work in the background.
                _cancellation = New CancellationTokenSource()
                Dim token = _cancellation.Token
                Dim runParallel = chkParallel.Checked

                ' When the task completes, update the UI.
                Task.Factory.StartNew(Sub()
                                          ' Create the new genetic algorithm.
                                          Dim ga = New TextMatchGeneticAlgorithm(runParallel, _targetText, settings)
                                          Dim bestGenome? As TextMatchGenome = Nothing
                                          _currentIteration = 1
                                          ' Iterate until a solution is found or until cancellation is requested.
                                          Do
                                              token.ThrowIfCancellationRequested()

                                              ' Move to the next generation.
                                              ga.MoveNext()
                                              ' If we've found the best solution thus far, update the UI.
                                              If bestGenome Is Nothing OrElse ga.CurrentBest.Fitness < bestGenome.Value.Fitness Then
                                                  bestGenome = ga.CurrentBest
                                                  _uiTasks.StartNew(Sub()
                                                                        txtBestMatch.Text = bestGenome.Value.Text
                                                                    End Sub)
                                                  ' If we've found the solution, bail.
                                                  If bestGenome.Value.Text = _targetText Then
                                                      Exit Do
                                                  End If
                                              End If
                                              _currentIteration += 1
                                          Loop
                                      End Sub, token).ContinueWith(Sub(t)
                                                                       timerElapsedTime.Stop()
                                                                       chkParallel.Visible = True
                                                                       btnRun.Text = "Start"
                                                                       _cancellation = Nothing

                                                                       Select Case t.Status
                                                                           Case TaskStatus.Faulted
                                                                               MsgBox(t.Exception.ToString(), MsgBoxStyle.Critical, "Error")
                                                                           Case TaskStatus.RanToCompletion
                                                                               txtBestMatch.BackColor = Color.LightGreen
                                                                               lblGenerations.BackColor = Color.LemonChiffon
                                                                       End Select

                                                                   End Sub, _uiTasks.Scheduler)

            Else
                _cancellation.Cancel()
            End If


        End Sub

        Private _startTime As DateTimeOffset = DateTimeOffset.MinValue
        Private _lastTime As DateTimeOffset = DateTimeOffset.MinValue

        Private Sub timerElapsedTime_Tick(ByVal sender As Object, ByVal e As EventArgs) Handles timerElapsedTime.Tick
            Dim now = DateTimeOffset.Now
            Dim elapsed = CInt(Fix((now - _startTime).TotalSeconds))

            lblElapsedTime.Text = elapsed.ToString()
            lblGenerations.Text = _currentIteration.ToString()

            If elapsed > 2 Then
                Dim diffSeconds = (now - _lastTime).TotalSeconds
                If diffSeconds > 0 Then
                    lblGenPerSec.Text = (CInt(Fix(_currentIteration / diffSeconds))).ToString()
                End If
            End If
        End Sub
    End Class
End Namespace
